Tests/SwiftParserTest/AttributeTests.swift

```diff
@@ -73,41 +73,55 @@ final class AttributeTests: XCTestCase {
   func testMissingClosingParenToAttribute() {
     assertParse(
       """
-      @_specializeℹ️(e1️⃣
+      @_specializeℹ️(1️⃣e
       """,
       diagnostics: [
         DiagnosticSpec(
-          message: "expected ':' in attribute argument",
-          fixIts: ["insert ':'"]
+          locationMarker: "1️⃣",
+          message: "expected argument for '@_specialize' attribute",
+          fixIts: ["insert attribute argument"]
         ),
         DiagnosticSpec(
+          locationMarker: "1️⃣",
           message: "expected ')' to end attribute",
           notes: [NoteSpec(message: "to match this opening '('")],
           fixIts: ["insert ')'"]
         ),
         DiagnosticSpec(
-          message: "expected declaration after attribute",
-          fixIts: ["insert declaration"]
+          locationMarker: "1️⃣",
+          message: "expected declaration and ';' after attribute",
+          fixIts: ["insert declaration and ';'"]
         ),
       ],
       fixedSource: """
-        @_specialize(e:) <#declaration#>
+        @_specialize() <#declaration#>; e
```

Comment:
I did a bit of digging for this issue.
What I can see is, that the current token is `e`  we don't consume it, as `e`  is not a label in `@_spacialize(`
We then try to find the precedence of the current token. `e` is an identifier, thus there are none.

Thats why we get this result I think
```
testMissingClosingParenToAttribute(): failed - Actual output (+) differed from expected output (-):
–@_specialize(e:) <#declaration#>
+@_specialize() <#declaration#>; e
```

The parser tries to parse the label `e` and don't know it, so we break out if the loop as you suggested.
So I think it's doing its work correctly (for now)
But it would be more nice if we did consume `e` as unknown label, and then apply a fixit that suggest to make one if the ones defined in the token set.

Reply:
Ah, what I missed here is that there is no closing `)` in the test case. If the input was `@_specialize(e)` (which is what I thought, I just didn’t check carefully), then `expect(.rightParen)` would eat the `e` as unexpected and recover to `)`. So, I agree that the current recovery is consistent and makes sense for now. 

Also, this test case is a little degenerate because it doesn’t have a declaration following the attribute. If you make it a little bit more realistic like

```swift
@_specialize(e
func foo() {}
```

then the fixed source is

```swift
@_specialize()e
func foo() {}
```

Putting the `e` outside the parentheses makes as much sense as putting it inside, so I think it’s fine and there’s nothing we need to do.